modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1245 @@
+      localThis.concurrentTasks.waitFor(0);
+    }
+
+    // Mark as frozen...
+    coforall loc in targetLocales do on loc {
+      var localThis = getPrivatizedThis;
+      localThis.frozenState.write(FREEZE_FROZEN);
+    }
+
+    return true;
+  }
+
+  /*
+    Unfreezes our state, allowing mutating operations; we wait on any ongoing
+    concurrent operations to allow them to finish.
+  */
+  proc unfreeze() : bool {
+    if BAG_NO_FREEZE then return false;
+    // Check if already unfrozen
+    if frozenState.read() == FREEZE_UNFROZEN {
+      return true;
+    }
+
+    // Mark as transient state...
+    coforall loc in targetLocales do on loc {
+      var localThis = getPrivatizedThis;
+      localThis.frozenState.write(FREEZE_MARKED);
+      localThis.concurrentTasks.waitFor(0);
+    }
+
+    // Mark as unfrozen...
+    coforall loc in targetLocales do on loc {
+      var localThis = getPrivatizedThis;
+      localThis.frozenState.write(FREEZE_UNFROZEN);
+    }
+
+    return true;
+  }
+
+  proc canFreeze() : bool {
+    return !BAG_NO_FREEZE;
+  }
+
+  /*
+    If we are currently frozen. If we are in the middle of a state transition, we
+    wait until it has completed.
+  */
+  proc isFrozen() : bool {
+    if BAG_NO_FREEZE then return false;
+
+    var state = frozenState.read();
+
+    // Current transitioning state, wait it out...
+    while state == FREEZE_MARKED {
+      chpl_task_yield();
+      state = frozenState.read();
+    }
+
+    return state == FREEZE_FROZEN;
+  }
+
+  /*
+    Clear all bags across all nodes in a best-effort approach. Elements added or
+    moved around from concurrent additions or removals (including work stealing)
+    may be missed while clearing.
+  */
+  proc clear() {
+    var localThis = getPrivatizedThis;
+    if !BAG_NO_FREEZE {
+      local {
+        localThis.concurrentTasks.add(1);
+
+        // Check if the queue is now 'immutable'.
+        if localThis.frozenState.read() > FREEZE_UNFROZEN {
+          localThis.concurrentTasks.sub(1);
+          return;
+        }
+      }
+    }
+
+    coforall loc in localThis.targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      forall segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+        if segment.acquireIfNonEmpty(STATUS_REMOVE) {
+          var block = segment.headBlock;
+          while block != nil {
+            var tmp = block;
+            block = block.next;
+            delete tmp;
+          }
+
+          segment.headBlock = nil;
+          segment.tailBlock = nil;
+          segment.nElems.write(0);
+          segment.releaseStatus();
+        }
+      }
+    }
+
+    localThis.concurrentTasks.sub(1);
+  }
+
+  /*
+    Triggers a more static approach to load balancing; note that this operation
+    is parallel-safe but is performed in parallel across all `targetLocales` and
+    should only be called at a point of synchronization.
+  */
+  proc balance() {

```

Comment:
Doing a coforall over segments is necessary in that we steal horizontally, we steal from all segments with the same segment index as ours. Since everything is going to be evenly locally distributed, then  this distribution can be kept and balanced across all horizontal segments.

The latter is definitely due to ease of implementation.